internal/frameworkprovider/slo_resource.go

```diff
@@ -0,0 +1,344 @@
+		moveSLOUpdateWarningFunc()
+		return
+	}
+}
+
+// Delete is called when the provider must delete the resource. Config
+// values may be read from the DeleteRequest.
+//
+// If execution completes without error, the framework will automatically
+// call DeleteResponse.State.RemoveResource(), so it can be omitted
+// from provider logic.
+func (s *SLOResource) Delete(ctx context.Context, req resource.DeleteRequest, resp *resource.DeleteResponse) {
+	var model SLOResourceModel
+	resp.Diagnostics.Append(req.State.Get(ctx, &model)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	diags := s.client.DeleteObject(ctx, manifest.KindSLO, model.Name, model.Project)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+}
+
+// ImportState is called when the provider must import the resource's state.
+func (s *SLOResource) ImportState(
+	ctx context.Context,
+	req resource.ImportStateRequest,
+	resp *resource.ImportStateResponse,
+) {
+	parts := strings.Split(req.ID, "/")
+	if len(parts) != 2 {
+		resp.Diagnostics.AddError(
+			"Invalid import ID",
+			"Expected ID to be in the format '<project-name>/<slo-name>'.",
+		)
+		return
+	}
+	model, diags := s.readResource(ctx, &SLOResourceModel{Name: parts[1], Project: parts[0]})
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	resp.Diagnostics.Append(resp.State.Set(ctx, model)...)
+}
+
+func (s *SLOResource) Configure(
+	_ context.Context,
+	req resource.ConfigureRequest,
+	resp *resource.ConfigureResponse,
+) {
+	if req.ProviderData == nil {
+		return
+	}
+	client, ok := req.ProviderData.(*sdkClient)
+	if !ok {
+		addInvalidSDKClientTypeDiag(&resp.Diagnostics, req.ProviderData)
+		return
+	}
+	s.client = client
+}
+
+func (s *SLOResource) applyResource(
+	ctx context.Context,
+	model *SLOResourceModel,
+) (*SLOResourceModel, diag.Diagnostics) {
+	slo := model.ToManifest()
+	diagnostics := s.client.ApplyObject(ctx, slo)
+	if diagnostics.HasError() {
+		return nil, diagnostics
+	}
+
+	// Read the SLO after creation to fetch the computed fields.
+	appliedModel, diagnostics := s.readResource(ctx, model)
+	if diagnostics.HasError() {
+		return nil, diagnostics
+	}
+	return appliedModel, diagnostics
+}
+
+// readResource reads the current state of the resource from the Nobl9 API.
+func (s *SLOResource) readResource(
+	ctx context.Context,
+	model *SLOResourceModel,
+) (*SLOResourceModel, diag.Diagnostics) {
+	slo, diagnostics := s.client.GetSLO(ctx, model.Name, model.Project)
+	if diagnostics.HasError() {
+		return nil, diagnostics
+	}
+	updatedModel := newSLOResourceConfigFromManifest(slo)
+	s.sortLists(model, updatedModel)
+	return updatedModel, diagnostics
+}
+
+// sortLists sorts lists returned by the API to ensure consistent ordering.
+func (s *SLOResource) sortLists(model, updatedModel *SLOResourceModel) {
+	updatedModel.Labels = sortLabels(model.Labels, updatedModel.Labels)
+	updatedModel.Objectives = sortListBasedOnReferenceList(
```

Comment:
why do we have to do it like that? why isn't a regular sort good enough?

also, shouldn't the equals func take project into consideration?

Reply:
note that previously, we had some of these blocks as sets, so it worked a bit different

I changed them to lists, after some consultations with the folks from Hashicorp, it turned out that using sets for such complex lists of objects is an anti-pattern and not recommended (to say the least), and yes, I've really tried to make it work with sets, but it was just pure hell

The thread I started with them is here: https://discuss.hashicorp.com/t/setnestedblock-with-custom-custom-type-for-the-nested-object-doesnt-seem-to-work-as-expceted/75907